Userland/Libraries/LibJS/Runtime/ArrayBufferPrototype.cpp

```diff
@@ -37,51 +38,96 @@ ArrayBufferPrototype::~ArrayBufferPrototype()
 // 25.1.5.3 ArrayBuffer.prototype.slice ( start, end ), https://tc39.es/ecma262/#sec-arraybuffer.prototype.slice
 JS_DEFINE_NATIVE_FUNCTION(ArrayBufferPrototype::slice)
 {
+    // 1. Let O be the this value.
+    // 2. Perform ? RequireInternalSlot(O, [[ArrayBufferData]]).
     auto* array_buffer_object = TRY(typed_this_value(global_object));
-    // FIXME: Check for shared buffer
+
+    // 3. If IsSharedArrayBuffer(O) is true, throw a TypeError exception.
+    // fixme: shared buffer
+
+    // 4. If IsDetachedBuffer(O) is true, throw a TypeError exception.
     if (array_buffer_object->is_detached())
         return vm.throw_completion<TypeError>(global_object, ErrorType::DetachedArrayBuffer);
 
+    // 5. Let len be O.[[ArrayBufferByteLength]].
     auto length = array_buffer_object->byte_length();
 
+    // 6. Let relativeStart be ? ToIntegerOrInfinity(start).
     auto relative_start = TRY(vm.argument(0).to_integer_or_infinity(global_object));
 
     double first;
-    if (relative_start < 0)
+    if (Value(relative_start).is_negative_infinity()) {
+        // 7. If relativeStart is -∞, let first be 0.
+        first = 0;
+    } else if (relative_start < 0) {
+        // 8. Else if relativeStart < 0, let first be max(len + relativeStart, 0).
         first = max(length + relative_start, 0.0);
-    else
+    } else {
+        // 9. Else, let first be min(relativeStart, len).
         first = min(relative_start, (double)length);
+    }
 
+    // 10. If end is undefined, let relativeEnd be len; else let relativeEnd be ? ToIntegerOrInfinity(end).
     auto relative_end = vm.argument(1).is_undefined() ? length : TRY(vm.argument(1).to_integer_or_infinity(global_object));
 
     double final;
-    if (relative_end < 0)
+    if (Value(relative_end).is_negative_infinity()) {
+        // 11. If relativeEnd is -∞, let final be 0.
+        final = 0;
+    } else if (relative_end < 0) {
+        // 12. Else if relativeEnd < 0, let final be max(len + relativeEnd, 0).
         final = max(length + relative_end, 0.0);
-    else
+    } else {
+        // 13. Else, let final be min(relativeEnd, len).
         final = min(relative_end, (double)length);
+    }
 
+    // 14. Let newLen be max(final - first, 0).
     auto new_length = max(final - first, 0.0);
 
+    // 15. Let ctor be ? SpeciesConstructor(O, %ArrayBuffer%).
     auto* constructor = TRY(species_constructor(global_object, *array_buffer_object, *global_object.array_buffer_constructor()));
+
+    // 16. Let new be ? Construct(ctor, « 𝔽(newLen) »).
     auto* new_array_buffer = TRY(construct(global_object, *constructor, Value(new_length)));
 
+    // 17. Perform ? RequireInternalSlot(new, [[ArrayBufferData]]).
+    // fixme: is this the exception we want here?
```

Comment:
yes?

Reply:
I had meant to provide a comment here, or on Discord on this, I got drawn away by something else after submitting the pr.

My point here is, in the spec both step 2 and 17 are performing `RequireInternalSlot(new, [[ArrayBufferData]])` - however are throwing two different exceptions here.
- Step 1 throws `vm.throw_completion<TypeError>(global_object, ErrorType::NotAnObjectOfType, PrototypeType::display_name())`
- Step 17 throws `vm.throw_completion<TypeError>(global_object, ErrorType::SpeciesConstructorDidNotCreate, "an ArrayBuffer")`

Reply:
They check slightly different things though :)

The first is for the this value, the second for the object created by the constructor returned from `SpeciesConstructor`, so I think it's ok to throw different errors.